### PR TITLE
Upgrade to Lucene 9.12.0

### PR DESCRIPTION
**Pull request by ChrisHegarty**
Upgrade to Lucene 9.12

This PR will be squash-merged, so as to simplify the backport to the 8.x branch.

### CODE REVIEW

**Comment by ChrisHegarty** [comment]
@elasticsearchmachine update branch

File: build.gradle [hunk 1]
```diff
@@ -241,6 +241,11 @@ if (project.gradle.startParameter.taskNames.any { it.startsWith("checkPart") ||
 
 subprojects { proj ->
   apply plugin: 'elasticsearch.base'
+
+  repositories {
+    // TODO: Temporary for Lucene RC builds. REMOVE
+    maven { url "https://dist.apache.org/repos/dist/dev/lucene/lucene-9.12.0-RC1-rev-f9cb943157870156af5f31cce315b95c08295c03/lucene/maven" }
+  }
```
Comment (ChrisHegarty) [inline build.gradle:248]: This will be reverted once 9.12 is released.

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine rerun elasticsearch-ci/part-2-arm

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
Lucene 9.12.0 has been released. This PR is ready of final review.  I cannot officially review in GH, but it LGTM.

**Comment by benwtrent** [comment]
My only slight concern is the flat vector formats. Now Lucene 9.12 treats them like exposed normal codecs we need to make sure there aren't any weird name conflicts, etc. with our flat formats.

I don't think there should be since ours have a different name. Just wanting to be sure.

File: server/src/main/java/org/elasticsearch/node/NodeConstruction.java [hunk 2]
```diff
@@ -538,6 +540,17 @@ private SettingsModule validateSettings(Settings envSettings, Settings settings,
         return settingsModule;
     }
 
+    @UpdateForV9
+    private static void addBwcSearchWorkerSettings(List<Setting<?>> additionalSettings) {
+        // TODO remove the below settings, they are unused and only here to enable BwC for deployments that still use them
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/node/NodeConstruction.java:545]: this may deserve some better annotation than a TODO, perhaps `@UpdateForV9`

**Review by javanna — APPROVED** [review]
I left a note on a TODO, I missed it when reviewing the corresponding PR. We can also address this post-merge. LGTM otherwise.

File: server/src/main/java/org/elasticsearch/node/NodeConstruction.java (hunk 2 above)
Comment (ChrisHegarty) [reply server/src/main/java/org/elasticsearch/node/NodeConstruction.java:546]: it already has @UpdateForV9, but maybe you added that.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/node/NodeConstruction.java:546]: oh boy, I have already done that then... but I did not see it when I reviewed. all good

**Pull request merged**